docs(blade): Tabs API Decisions (#1693)
razorpay/blade: packages/blade/src/components/Tabs/_decisions/decisions.md

# Tabs API Decisions <!-- omit in toc -->

A tab is a navigation component used in the interface to switch between different views in the same context. Tabs are contextual to the section or the page and are triggered by user interaction.

<img src="./tabs-thumbnail.png" width="380" alt="Tabs thumbnail" />

- [Design](#design)
- [Features](#features)
- [Anatomy](#anatomy)
- [`Tabs` API](#tabs-api)
- [`TabsItem` API](#tabsitem-api)
- [`TabsPanel` API](#tabspanel-api)
  - [Examples:](#examples)
- [Basic](#basic)
  - [Controlled, Uncontrolled](#controlled-uncontrolled)
- [Filled Variant](#filled-variant)
- [With Tooltip](#with-tooltip)
- [Motion](#motion)
- [Accessibility](#accessibility)
- [Open Questions](#open-questions)

## Design

[Figma Link](https://www.figma.com/file/LSG77hEeVYDk7j7WV7OMJE/Blade-DSL---Components-Guideline?node-id=3733%3A8981&mode=dev) to all variants of the Tabs component

## Features

- Controlled and Uncontrolled
- Filled and Bordered variants
- Horizontal and Vertical orientation
- Medium and Large size

## Anatomy

<img src="./tabs-anatomy.png" width="100%" alt="Tabs thumbnail" />

> ```jsx

Reviewer: How about
```jsx
<Tabs>
  <TabList>
     <TabItem>
     </TabItem>
  </TabList>

  <TabPanel>
  </TabPanel>
</Tabs>

```

Reviewer: TabPanel still seems very decoupled from TabItem, maybe we can nest them together somehow?

Reviewer: How about something straightforward? What is the downside of this?
```jsx
<Tabs>
     <Tab value={} leading={} trailing={} title={}>
        {...the jsx content for the content of the tab}
     </Tab>
</Tabs>
```

Author: @chaitanyadeorukhkar while the API that you proposed looks much compact and straightforward there are major downsides to this API

### 1. Quite tricky/hack to implement

With the API that you proposed it will be quite tricky to map the JSX to the HTML structure that a typical tab requires. 

A tab's HTML structure is as follows

```html
<div>
  <div role="tablist">
    <button role="tab">tab item 1</button>
    <button role="tab">tab item 2</button>
    <button role="tab">tab item 3</button>
  </div>
  
  <div>
    <section role="tabpanel" />
    <section role="tabpanel" />
    <section role="tabpanel" />
  </div>
</div>
```

We need to achieve the above HTML structure, and to do this with the API that you proposed will get very hacky since now we need to use React.cloneElement and save those JSX nodes in state and render them in appropriate place. 

```html
<Tabs>
      // this will be the button with role="tab", need to React.cloneElement and move it to the right place, and while placing also need to make sure it has a parent of role="tablist"
     <Tab value={} leading={} trailing={} title={}>  
        // this is role="tabpanel" need to React.cloneElement and move it to the right place
        {...the jsx content for the content of the tab} 
     </Tab>
</Tabs>
```

### 2. The JSX Doesn't map well with the visual hierarchy or the HTML structure

If you notice the HTML structure pointed out in point #1 you'll see that it's 1:1 with the JSX API that i'm proposing. 

The current API that is proposed also maps 1:1 with the visual hierarchy of the component, but the second one doesn't map very well.  

![Frame 1 (11)](https://github.com/razorpay/blade/assets/35374649/9058d53f-aa5b-4a45-9111-5bfbdafa8f27)

### 3. Extensibility

The current API is also much more easier to extend in case any future usecase arises, for example let's say we want to use these tabs like a segmented control (which works like radio buttons but doesn't associate with any "tabpanels" we can simply remove the tabpanels and change the underlying html structure to acomodate for any future usecases. 

Check how reshaped is allowing this [extensibility](https://reshaped.so/content/docs/components/tabs#using-in-forms) - if you pass `name` prop it will convert the tabs to work like a Radio. 

### 4. API Familiarity 

The current API that is proposed is also much more familiar and is similar to how many other design systems are approaching compared to the second one: 

- [Reshaped](https://reshaped.so/content/docs/components/tabs)
- [Chakra](https://chakra-ui.com/docs/components/tabs/usage)
- [MUI](https://mui.com/material-ui/react-tabs/#experimental-api)
- [Atlassian](https://atlassian.design/components/tabs/examples)
- [RadixUI](https://www.radix-ui.com/primitives/docs/components/tabs)
- [HeadlessUI](https://headlessui.com/react/tabs)

Author: Added

Author: https://github.com/razorpay/blade/pull/1693#discussion_r1345552656

Also changed the naming of the current components to not use the plural suffix. instead of TabsItem -> TabItem

  - [3. Extensibility](#3-extensibility)
  - [4. API Familiarity](#4-api-familiarity)
- [Motion](#motion)
- [Accessibility](#accessibility)
- [Open Questions](#open-questions)

## Design

[Figma Link](https://www.figma.com/file/LSG77hEeVYDk7j7WV7OMJE/Blade-DSL---Components-Guideline?node-id=3733%3A8981&mode=dev) to all variants of the Tabs component

## Features

- Controlled and Uncontrolled
- Filled and Bordered variants
- Horizontal and Vertical orientation
- Medium and Large size

## Anatomy

<img src="./tabs-anatomy.png" width="100%" alt="Tabs thumbnail" />

```jsx
<Tabs>
  <TabList>
    <TabItem />
  </TabList>
  <TabPanel />
</Tabs>
```

## `Tabs` API

```ts
type TabsProps = {
  /**
   * The content of the component, accepts `TabList` and `TabPanel` components.
   */
  children: React.ReactNode;
  /**
   * The value of the selected tab, If set the component will be controlled.
   */
  value?: string;
  /**
   * The default value of the selected tab, in case the Tabs component is uncontrolled.
   */
  defaultValue?: string;
  /**
   * Callback fired when the value changes.
   */
  onChange?: (value: string) => void;
  /**
   * The orientation of the tabs.
   */

>   orientation?: 'horizontal' | 'vertical';

Reviewer 2: what if we call this as `direction: row | column`?

Author: orientation I think sounds more natural here, direction might be confused with the flex property or the CSS direction attribute. (plus could cause issues if we have to support rtl, ltr)

  /**
   * The size of the tabs.
   */
  size?: 'medium' | 'large';
  /**
   * The variant of the tabs.
   */
  variant?: 'bordered' | 'filled';
  /**
   * If `true`, the TabItems will grow to use all the available space.
   */

>   autoWidth?: boolean;

Reviewer 3: Okay. So in this case the `TabItems` would stretch to occupy the whole `Tabs`, right?

>   autoWidth?: boolean;

Author: Yeah check the design [here](https://www.figma.com/file/LSG77hEeVYDk7j7WV7OMJE/Blade-DSL---Components-Guideline?node-id=3733%3A10954&mode=dev)

<img width="648" alt="image" src="https://github.com/razorpay/blade/assets/35374649/a8412e90-5e4b-4487-8f5f-c86c5e9177f6">

>   autoWidth?: boolean;

Reviewer 2: autoWidth is confusing too. because i'm unable to understand what will it do? will it adjust the width of the tab or tablist or tab panel as it affects the width of the tabList. can we name it something like `isFullWidthTabList`?

>   autoWidth?: boolean;

Author: isFullWidthTabList, too verbose. 

I've seen design systems use "isFitted" prop. which might be better?

>   autoWidth?: boolean;

Reviewer 2: again adds a learning curve. i'm okay with verbose name as far as it conveys the purpose without me fiddling around and guessing what will it do.

};
```

## `TabItem` API

```ts
type TabItemProps = {
  /**
   * The label of the tab item.
   */

>   children: string;

Reviewer 2: shouldn't this be `StringChildrenType`?

Author: Yeah sure, we can do this to enable interpolation.

Author: Updated to use React.ReactNode.

  /**
   * The value of the tab item.
   */
  value: string;
  /**
   * Leading element of the tab item.

>    * Can be used to render an icon.

Reviewer 2: are we putting a guardrail for this?

Author: Yes.

   */
  leading?: React.ReactNode;
  /**
   * Trailing element of the tab item.
   * Can be used to render an badge/counter.
   */
  trailing?: React.ReactNode;
  /**
   * If `true`, the tab item will be disabled.
   */
  isDisabled?: boolean;
};
```

## `TabPanel` API

> 

Reviewer 4: Can we employ lazy loading for various tab panels? Maybe an example will help. Also will the tab panels be unmounted and removed from DOM when unselected?

Author: 🤔 Hmm. 

> Could probably expose a lazy prop
When you say "lazy loading tabs" do you mean to have the Tab's content be a Suspense component? Or do you mean a "lazy" prop which will control the mounted behaviour of the tabs, eg: lazy=true ---> unmounts on selection change, lazy=false ---> doesn't unmount just hides via display: none.

> Also will the tab panels be unmounted and removed from DOM when unselected?
right now we unmount unselected tabs.

Reviewer 2: >right now we unmount unselected tabs.

this could be trickier because in some product cases we might need to retain the state of the children(for eg checkbox or an accordion in the previous selected tab)

Author: > this could be trickier because in some product cases we might need to retain the state of the children(for eg checkbox or an accordion in the previous selected tab)

This can be solved, by adding lazy prop, on react-native side, [react-native-tab-view](https://reactnavigation.org/docs/tab-view/) also provides a similar API. Should be simple enough to implement.

```ts
type TabPanelProps = {
  /**
   * The value of the tab panel. This will be used to match the selected tab.
   */
  value: string;
  /**
   * The content of the tab panel.
   */
  children: React.ReactNode;
};

> ```

Reviewer 2: add `TabList` props?

Author: TabList doesn't have any props, it's just used for structuring and internal logic management.

Reviewer 2: yes so it'll least have children right? better to add in the spec to avoid any confusion

> ### Examples:

Reviewer 4: Examples missing or wrong heading level?

## Basic

```js
<Tabs>
  <TabList>
    <TabItem value="payments">Payments</TabItem>
    <TabItem value="refunds">Refunds</TabItem>
    <TabItem value="disputes">Disputes</TabItem>
    <TabItem value="settlements">Settlements</TabItem>
  </TabList>

>   <TabPanel value="payments">

Reviewer 4: One con of this API is I'm not sure if we can make `value` type safe? For example, easy to mis-spell `value` in tab item and panel

Author: Yeah I think that we can't do anything. 

But should be fine, either way you should not have more than 7 or 8 tabs (maybe max 10), so don't foresee any maintainability issues. 

People can also do this

```
const paymentTabValue = "payments"

<TabItem value={paymentTabValue} />
<TabContent value={paymentTabValue} />
```

    <Payments />
  </TabPanel>

>   <TabPanel value="refunds">

Reviewer 2: do we give the styling of tabpanel like a card or something or just an empty box and consumer decides the styling of panel?

Author: It's an empty box, no padding, no margin, no styling.

    <Refunds />
  </TabPanel>
  <TabPanel value="disputes">
    <Disputes />
  </TabPanel>
  <TabPanel value="settlements">
    <Settlements />
  </TabPanel>
</Tabs>
```

> <img src="./example-1.png" width="380" alt="Tabs Example 1" />

Reviewer 2: can you attach an example that is the output of the above code? this seems a made screenshot from existing page?

Author: No it's not from an existing page, the screenshot is from the guidelines example RK added.

  return (
    <Tabs defaultValue="refunds">
      <TabList>
        <TabItem value="payments">Payments</TabItem>
        <TabItem value="refunds">Refunds</TabItem>
      </TabList>

      <TabPanel value="payments">
        <Payments />
      </TabPanel>
      <TabPanel value="refunds">
        <Refunds />
      </TabPanel>
    </Tabs>
  );
};
```

## Vertical Orientation

```jsx
<Tabs orientation="vertical">
  <TabList>
    <TabItem value="all-expenses" leading={RupeeIcon}>
      All Expenses
    </TabItem>
    <TabItem
      value="categories"
      leading={EditIcon}
      trailing={<Counter value={26} variant="notice" />}
    >
      Categorise
    </TabItem>
    <TabItem
      value="review"
      leading={TransactionIcon}
      trailing={<Counter value={99} variant="information" />}
    >
      Review
    </TabItem>
    <TabItem value="synced" leading={CheckCircleIcon}>
      Synced
    </TabItem>
    <TabItem
      value="synced-fail"
      leading={WarningIcon}
      trailing={<Counter value={3} variant="negative" />}
    >
      Synced Failed
    </TabItem>
    <TabItem value="excluded" leading={SlashIcon}>
      Excluded
    </TabItem>
  </TabList>
  ...
</Tabs>
```

> <img src="./example-4.png" width="380" alt="Tabs Example 2" />

Reviewer 2: will the UI look like this exactly?

Author: Yeah mostly, omitting whatever inside the <TabPanels \/>

## Filled Variant

```jsx

> <Tabs defaultValue="refunds">

Reviewer 2: update this. typo here. it doesn't has filled variant and the default value is something else

Author: Updated.

  <TabList>
    <TabItem value="week">Week</TabItem>
    <TabItem value="month">Month</TabItem>
    <TabItem value="custom">Custom</TabItem>
  </TabList>

  <TabPanel value="week">
    <WeekGraph />
  </TabPanel>
  <TabPanel value="month">
    <MonthGraph />
  </TabPanel>
  <TabPanel value="custom">
    <CustomGraph />
  </TabPanel>
</Tabs>
```

<img src="./example-2.png" width="380" alt="Tabs Example 2" />

## With Tooltip

```jsx
<Tabs>
  <TabList>
    <TabItem value="payments">Payments</TabItem>
    <TabItem value="refunds">

>       <Tooltip content="Return (money) in restitution, repayment, or balancing of accounts.">

Reviewer 2: the props table above says that the children will be string? but we are passing JSX here. will this work?

Author: Oh wait right, we have to keep this React.ReactNode then to accommodate tooltip usecases.

Author: Updated.

        <TooltipInteractiveWrapper>Refunds</TooltipInteractiveWrapper>
      </Tooltip>
    </TabItem>
    <TabItem value="disputes">Disputes</TabItem>
    <TabItem value="settlements">Settlements</TabItem>
  </TabList>
  ...
</Tabs>
```

<img src="./example-3.png" width="380" alt="Tabs Example 3" />

## Alternative APIs

We [discussed](https://github.com/razorpay/blade/pull/1693#discussion_r1345677600) another potential API: 

```jsx
<Tabs>
  <Tab title="Payments" value="" leading={} trailing={}>
    <Text>Tab Panel Content For Payments</Text>
  </Tab>
  <Tab title="Refunds" value="" leading={} trailing={}>
    <Text>Tab Panel Content For Refunds</Text>
  </Tab>
  <Tab title="Disputes" value="" leading={} trailing={}>
    <Text>Tab Panel Content For Disputes</Text>
  </Tab>
</Tabs>
```

We decided not to go with this API because of these downsides: 

#### 1. Quite tricky/hack to implement

> With the API that you proposed it will be quite tricky to map the JSX to the HTML structure that a typical tab requires. 

Reviewer 2: ```suggestion
With this API it will be quite tricky to map the JSX to the HTML structure that a typical tab requires. 
```

A tab's HTML structure is as follows

```html
<div>
  <div role="tablist">
    <button role="tab">tab item 1</button>
    <button role="tab">tab item 2</button>
    <button role="tab">tab item 3</button>
  </div>
  
  <div>
    <section role="tabpanel"></section>
    <section role="tabpanel"></section>
    <section role="tabpanel"></section>
  </div>
</div>
```

> We need to achieve the above HTML structure, and to do this with the API that you proposed will get very hacky since now we need to use React.cloneElement and save those JSX nodes in state and render them in appropriate place. 

Reviewer 2: ```suggestion
We need to achieve the above HTML structure, and to do this with the above API it will get very hacky since now we need to use React.cloneElement and save those JSX nodes in state and render them in appropriate place. 
```

#### 3. Extensibility

The current API is also much more easier to extend in case any future usecase arises, for example let's say we want to use these tabs like a segmented control (which works like radio buttons but doesn't associate with any "tabpanels" we can simply remove the tabpanels and change the underlying html structure to accommodate for any future usecases. 

Check how reshaped is allowing this [extensibility](https://reshaped.so/content/docs/components/tabs#using-in-forms) - if you pass `name` prop it will convert the tabs to work like a Radio. 

#### 4. API Familiarity 

The current API that is proposed is also much more familiar and is similar to how many other design systems are approaching compared to the second one: 

- [Reshaped](https://reshaped.so/content/docs/components/tabs)
- [Chakra](https://chakra-ui.com/docs/components/tabs/usage)
- [MUI](https://mui.com/material-ui/react-tabs/#experimental-api)
- [Atlassian](https://atlassian.design/components/tabs/examples)
- [RadixUI](https://www.radix-ui.com/primitives/docs/components/tabs)
- [HeadlessUI](https://headlessui.com/react/tabs)

> ## Motion

Reviewer 2: put the motion video of tabs switching for both vertical+horizontal orientation and bordered and filled variant

Author: There are many variants, attached the motion prototype instead. People can open the prototype and inspect the motion.

Reviewer 2: can you record a tiny video from the prototype and it as a sample here? want it to act as a cue for readers to open the prorotype

Reviewer 2: @anuraghazra bumping this up. link will easy be missed and rarely people will actually open it

Author: Added

> Check the motion [documentation](https://www.figma.com/proto/LSG77hEeVYDk7j7WV7OMJE/Blade-DSL---Components-Guideline?type=design&node-id=4005-31894&t=b3OlPHxPuFH1BAIZ-0&scaling=min-zoom&page-id=3698%3A13772) here.

Reviewer 2: ```suggestion
Check the motion [prototype](https://www.figma.com/proto/LSG77hEeVYDk7j7WV7OMJE/Blade-DSL---Components-Guideline?type=design&node-id=4005-31894&t=b3OlPHxPuFH1BAIZ-0&scaling=min-zoom&page-id=3698%3A13772) here.
```

## Accessibility

Tabs will follow the [WAI-ARIA Tabs](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/) pattern.

- [Keyboard navigation](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/#keyboardinteraction)

> ## Open Questions

Reviewer 5: Are the tabs SEO-friendly by default, or do we need to make any adjustments?
Will the tab panels be server-rendered to facilitate SEO?

Author: By default all TabPanels be mounted on DOM and we will show/hide them via CSS to ensure SEO. 

And additionally, we will also introduce a `isLazy` prop which if present, will unmount the TabPanels from the DOM, could be useful if imagine inside your TabPanel you are doing API requests and don't want to mount/init them.
